src/Utils/StrictPlus.php

```diff
@@ -0,0 +1,126 @@
+<?php
+
+/**
+ * This file is part of the Nette Framework (https://nette.org)
+ * Copyright (c) 2004 David Grudl (https://davidgrudl.com)
+ */
+
+namespace Nette;
+
+use Nette\Utils\Callback;
+use Nette\Utils\ObjectMixin;
+
+
+/**
+ * Strict class for better experience.
+ * - 'did you mean' hints
+ * - access to undeclared members throws exceptions
+ * - support for @property annotations
+ * - support for calling event handlers stored in $onEvent via onEvent()
+ */
+trait StrictPlus
+{
+
+	/**
+	 * @return mixed
+	 * @throws MemberAccessException
+	 */
+	public function __call($name, $args)
+	{
+		$class = get_class($this);
+
+		if (ObjectMixin::hasProperty($class, $name) === 'event') { // calling event handlers
+			if (is_array($this->$name) || $this->$name instanceof \Traversable) {
+				foreach ($this->$name as $handler) {
+					Callback::invokeArgs($handler, $args);
+				}
+			} elseif ($this->$name !== NULL) {
+				throw new UnexpectedValueException("Property $class::$$name must be array or NULL, " . gettype($this->$name) . ' given.');
+			}
+
+		} else {
+			ObjectMixin::strictCall($class, $name);
+		}
+	}
+
+
+	/**
+	 * @return void
+	 * @throws MemberAccessException
+	 */
+	public static function __callStatic($name, $args)
+	{
+		ObjectMixin::strictStaticCall(get_called_class(), $name);
+	}
+
+
+	/**
+	 * @return mixed   property value
+	 * @throws MemberAccessException if the property is not defined.
+	 */
+	public function & __get($name)
+	{
+		$class = get_class($this);
+
+		if ($prop = ObjectMixin::getMagicProperties($class)[$name] ?? NULL) { // property getter
+			if (!($prop & 0b0001)) {
+				throw new MemberAccessException("Cannot read a write-only property $class::\$$name.");
+			}
+			$m = ($prop & 0b0010 ? 'get' : 'is') . $name;
+			if ($prop & 0b0100) {
+				return $this->$m();
+			} else {
+				$val = $this->$m();
+				return $val;
+			}
+		} else {
```

Comment:
What about

~~~php
$class = get_class($this);
$prop = ObjectMixin::getMagicProperties($class)[$name] ?? NULL;

if ($prop & 0b0001) {
    $m = ($prop & 0b0010 ? 'get' : 'is') . $name;
    if ($prop & 0b0100) {
        return $this->$m();
    } else {
        $val = $this->$m();
        return $val;
    }

} elseif ($prop !== NULL) {
    throw new MemberAccessException("Cannot read a write-only property $class::\$$name.");

} else {
    ObjectMixin::strictGet($class, $name);
}
~~~